[Perf improvement] Implementing memoization for sanitizeNamespace and replaceGenericParameters funcs.

rafael-assis opened the pull request:
Hi @nalexn. Picking up were we left off on #263, we have identified another opportunity for performance improvements in ViewInspector. @bachand and I discussed it and we'd love to get your thoughts on it.
Thank you so much for your time! 🙏 

## Context: performance improvements under larger SwiftUI View hierarchies

We have profiled and traced some of our slowest tests in Instruments showed that the 2 funcs defined in the `String` extension were the top offenders in terms of running time.

- `String.replacingGenericParameters(_:)`
- `String.sanitizingNamespace()`

The combined time of all calls to both of these functions accounted for almost 50% of the entire test running time:
| Weight(s) | Weight(%) | Self Weight | Symbol Name
| --- | --- | --- | --- |
|28.26 s | 100.0%  |	0 s | UnitTestHost (16171)
8.08 s |28.5% | 8.08 s | String.replacingGenericParameters(_:)
5.80 s | 20.5%	 | 5.80 s | String.sanitizingNamespace()

![Instruments_trace](https://github.com/nalexn/ViewInspector/assets/2278022/85f30dc5-f341-4f75-83f0-b73732ad2ec7)

It was already known that these 2 functions are called thousands of times for a single test case in our codebase.
This fact motivated an investigation on how many times these 2 funcs are called for the exact same input in order to inform a potential decision to memoize the calls. 

A histogram of the calls to `sanitizingNamespace` for each input was built from a single feature integration test run.
We found that for some types, the func `sanitizingNamespace` thousands of times for them. For example:

```
    - key : "SwiftUI.Environment<Swift.Optional<Swift.AnyHashable>>"
    - value : 3640

    - key : "SwiftUICoreUI.ScrollEventDispatch"
    - value : 144817

    - key : "Swift.Bool"
    - value : 3906

    - key : "SwiftUI.Environment<PrimitiveTypesCoreUI.AdaptiveTraitCollection>"
    - value : 19052
 
    - key : "(PrimitiveTypesCoreUI.InteractivityState) -> SwiftUI.ModifiedContent<SwiftUI._ViewModifier_Content<PrimitiveTypesCoreUI.Interactive<PrimitiveTypesCoreUI.TextStyleViewModifier>>, PrimitiveTypesCoreUI.TextStyleViewModifier>"
    - value : 17912
```

## Changes

It was then clear that caching the output of these expensive string manipulation operations would be worth it.
This is exactly what theses changes do.
It's a simple memoization which stores the computed value to be returned directly instead of recalculating them repeatedly.

## Testing

-  Ran tests at Airbnb that use ViewInspector

Local runs of the the test used to profile ViewInspector's execution time showed a **substantial improvement of running time. During the 2 Unit Tests CI runs of the changes in this PR, we noticed an **improvement of about 86% (19 seconds)**. That particular test took ~22 seconds to run before the optimizations and  ~3 seconds** after the optimizations:

Although the unit tests in ViewInspector are more lightweight compared to the tests we run at Airbnb, we could also see a good improvement of the overall running time of all tests of the library:

| Before | After |
| --- | --- |
| <img width="1072" alt="ViewInspector_perf_before" src="https://github.com/nalexn/ViewInspector/assets/2278022/179c57ed-8529-4643-9059-3794448d85cc"> | <img width="1072" alt="ViewInspector_perf_after" src="https://github.com/nalexn/ViewInspector/assets/2278022/3b50b7b6-5a12-4a01-89d5-5b7a8e450a17"> |

## Please review:

@nalexn 
@michael-bachand

rafaelassis-sys committed:
Implementing memoization for sanitizingNamespace and replaceGenericParameters funcs.

rafael-assis commented on Sources/ViewInspector/Inspector.swift:
```diff
@@ -1,18 +1,115 @@
+    static func sanitizeNamespace(ofTypeName typeName: String) -> String {
+        var str = typeName
+
+        if let sanitized = sanitizedNamespacesCache[typeName] {
```
I moved the exact implementation from the `String` extension to here and added the memoization calls here and in line 26.

rafael-assis commented on Sources/ViewInspector/Inspector.swift:
```diff
@@ -1,18 +1,115 @@
+        if balance == 0 {
+            return String(typeName[..<start]) +
+            replacement +
+            replaceGenericParameters(inTypeName: String(typeName[current...]),
```
We still call into the memoized func to potentially leverage from memoized subtypes.

bachand replied:
Smart!

rafael-assis commented on Sources/ViewInspector/Inspector.swift:
```diff
@@ -1,18 +1,115 @@
+    private static var sanitizedNamespacesCache: [String: String] = [:]
+
+    // swiftlint:disable force_try
+    private static let sanitizeNamespaceRegex = try! NSRegularExpression(pattern: "\\.\\(unknown context at ..........\\)")
```
We thought it'd be better to keep this regex as a constant instead of recreating it every time `sanitizeNamespace` is called.

rafael-assis commented on Sources/ViewInspector/Inspector.swift:
```diff
@@ -310,9 +361,9 @@ internal extension Inspector {
+
     static func guardType(value: Any, namespacedPrefixes: [String], inspectionCall: String) throws {
-        
+
```
@nalexn do you have a preference regarding the style of tabs / spaces / empty lines?
I can revert those to the original state if you prefer. Let me know.

nalexn replied:
Nah, I like this more. [I've googled](https://stackoverflow.com/a/72840452/2923345) how to tweak the Xcode settings to not insert spaces

rafael-assis replied:
Sounds good! I like this better too!

bachand replied:
@nalexn in case it's helpful, the [Airbnb style guide has a script](https://github.com/airbnb/swift#xcode-formatting) that you can run that enables various editor settings like this one that we find helpful in our large project.

nalexn commented:
That is a substantial performance improvement, good job! Before merging this in, I suggest an additional slight refactor of `sanitizeNamespace` in [this fasion](https://github.com/nalexn/ViewInspector/issues/268#issuecomment-1806104270).
The part where it removes `"(extension in ...)"` isn't yet confirmed to be helpful, but it'd be best to make sure your final version supports adding more patterns for prefix removal

rafaelassis-sys committed:
sanitizeNamespace: Support for multiple patterns

rafael-assis commented:
> That is a substantial performance improvement, good job! Before merging this in, I suggest an additional slight refactor of `sanitizeNamespace` in [this fasion](https://github.com/nalexn/ViewInspector/issues/268#issuecomment-1806104270). The part where it removes `"(extension in ...)"` isn't yet confirmed to be helpful, but it'd be best to make sure your final version supports adding more patterns for prefix removal

Thank you for the super quick turnaround, @nalexn! 🙏 
I have updated  `sanitizeNamespace` to support multiple patterns as you suggested.

I left the new pattern disabled for now with a comment. Let me know what you think of the approach.

rafael-assis commented on Sources/ViewInspector/Inspector.swift:
```diff
@@ -1,18 +1,125 @@
+
+    private static let sanitizeNamespaceRegex = {
+        // swiftlint:disable force_try
+        try! NSRegularExpression(pattern: sanitizeNamespacePatterns.joined(separator: "|"))
```
I believe having multiple patterns joined by an [OR operator](https://developer.apple.com/documentation/foundation/nsregularexpression#1661082) (`|`) might be faster than running the `stringByReplacingMatches` for each of them. Since we're not looking to replace different patterns with different strings, this seemed like the most performant approach.

bachand replied:
This makes sense, though I wonder if we want to wrap patterns in `(` and `)` before OR'ing them together to ensure that each pattern is considered atomically?

rafael-assis replied:
I believe the fact that we add an `|` already treats both sides of it as individual patterns to be matched. But I don't think it's an issue adding them, and it makes it more clear too, so I added them. 👍

rafael-assis commented on Sources/ViewInspector/Inspector.swift:
```diff
@@ -1,18 +1,125 @@
+        // This pattern may be helpful to solve issue #268.
+        // It will remain disabled until it is confirmed.
+        // https://github.com/nalexn/ViewInspector/issues/268
+        //"\\(extension in [a-zA-Z0-9]*\\)\\:",
```
@nalexn I left this disabled since #268 is not yet resolved with a conclusion that this additional pattern helps.
I also left a comment for context.

bachand commented on Sources/ViewInspector/Inspector.swift:
```diff
@@ -1,18 +1,125 @@
+        return typeName
+    }
+
+    private static var replacedGenericParametersCache: [String : [String : String]] = [:]
+    private static var sanitizedNamespacesCache: [String: String] = [:]
```
This strategy for caching is **not** thread-safe. We think that's OK (i.e. thread safety isn't necessary) but I wanted to flag it to @nalexn out of an abundance of caution 👍

rafael-assis replied:
Thanks for mentioning that @bachand! 
I missed this comment and did want to flag it to @nalexn as well! 👍

nalexn replied:
Yeah, no need to solve a problem that may not even exist. I'm 99% sure it should be fine without synchronization, we can add it later if needed

bachand commented on Sources/ViewInspector/Inspector.swift:
```diff
@@ -1,18 +1,125 @@
+
+    /// Removes the "(unknown context at <memory_address>)" portion of a type name.
+    /// Calls to this method are memoized and retained for the lifetime of the program.
+    /// - Parameter typeName: The raw type name. (e.g. ``SomeTypeName.(unknown context at $138b3290c).SomePropertyName``)
+    /// - Returns: The sanitized type name. (e.g. ``SomeTypeName.SomePropertyName``)
```
```suggestion
    /// - Parameter typeName: The raw type name. (e.g. `SomeTypeName.(unknown context at $138b3290c).SomePropertyName`)
    /// - Returns: The sanitized type name. (e.g. `SomeTypeName.SomePropertyName`)
```

Nit: My understanding is that the double backticks are used to create links to other types when documentation comments are displayed in Xcode. Since these are example types, I think single backticks are more appropriate.

rafael-assis replied:
Makes sense! I'll update to single backticks! 👍

bachand commented on Sources/ViewInspector/Inspector.swift:
```diff
@@ -1,18 +1,125 @@
+    /// - Parameters:
+    ///   - typeName: The original type name. (e.g. ``SomeTypeName<SomeGenericTypeParameter>``)
+    ///   - replacement: The string to replace the generic parameters with. (e.g. "<EmptyView>")
+    /// - Returns: The type name with its generic parameters replaced. (e.g. ``SomeTypeName<EmptyView>``)
```
```suggestion
    ///   - typeName: The original type name. (e.g. `SomeTypeName<SomeGenericTypeParameter>`)
    ///   - replacement: The string to replace the generic parameters with. (e.g. "<EmptyView>")
    /// - Returns: The type name with its generic parameters replaced. (e.g. `SomeTypeName<EmptyView>`)
```

Nit: See similar comment above for rationale.

bachand reviewed (approved):
Great work @rafael-assis and thanks for the quick review @nalexn ! 🙏

rafaelassis-sys committed:
Updating `` with ` in comments, enclosing patterns between parenthesis.

rafael-assis commented:
> That is a substantial performance improvement, good job! Before merging this in, I suggest an additional slight refactor of sanitizeNamespace in https://github.com/nalexn/ViewInspector/issues/268#issuecomment-1806104270.
The part where it removes "(extension in ...)" isn't yet confirmed to be helpful, but it'd be best to make sure your final version supports adding more patterns for prefix removal

Hi @nalexn. Do you think the changes I made fully addressed your suggestions?
Let me know if you need me to make any further adjustment before you merge this PR.

Thank you!

nalexn commented:
Merged. Thank you @rafael-assis and @bachand !
